operator/pkg/ciliumendpointbatch/manager.go

```diff
@@ -0,0 +1,337 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package ciliumendpointbatch
+
+import (
+	"fmt"
+	"math/rand"
+	"time"
+
+	meta_v1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	"github.com/cilium/cilium/pkg/annotation"
+	cilium_v2 "github.com/cilium/cilium/pkg/k8s/apis/cilium.io/v2alpha1"
+	"github.com/cilium/cilium/pkg/lock"
+)
+
+var (
+	// sequentialLetters contains lower case alphabets without vowels and few numbers.
+	// skipped vowels and numbers [0, 1] to avoid generating controversial names.
+	sequentialLetters = []rune("bcdfghjklmnpqrstvwxyz2456789")
+)
+
+// ceBatch holds local copy of CiliumEndpointBatch.
+type ceBatch struct {
+	ceb          *cilium_v2.CiliumEndpointBatch
+	backendMutex lock.RWMutex
+}
+
+// cebManager
+type cebManager interface {
+	InsertCepInCache(cep *cilium_v2.CoreCiliumEndpoint) (string, error)
+	RemoveCepFromCache(cep *cilium_v2.CoreCiliumEndpoint) error
+	getCebFromCache(cebName string) (*cilium_v2.CiliumEndpointBatch, error)
+	findCeb(cep *cilium_v2.CoreCiliumEndpoint) (*ceBatch, bool)
+	updateCebInCache(ceb *cilium_v2.CiliumEndpointBatch, deepCopy bool)
+	deleteCebFromCache(cebName string)
+	createCeb(name string) *ceBatch
+	getCepCount() int
+	getCebCount() int
+	getAllCeps() map[string]*cilium_v2.CoreCiliumEndpoint
+}
+
+// Implementation of FirstComeFirstServe batching mode. If new CEP is inserted,
+// then the CEP is queued in any one of the available CEB. CEPs are inserted into
+// CEBs without any preference or any priority.
+type cebManagerFcfs struct {
+
+	// cacheCepMap is used to map cepName to cebName
+	cacheCepMap *cepToCebMapping
+
+	// cacheCeb is used to map cebName to ceBatch object.
+	cacheCeb map[string]*ceBatch
+
+	// Aggregator is an interface, as CEPs are inserted or removed in a CEB,
+	// this would keep track of changes happen to a CEB and resolve them into
+	// a single actionable change.
+	cebsToSync *aggregator
+
+	// maxCepsInCeb is the maximum number of CiliumCoreEndpoint(s) packed in
+	// a CiliumEndpointBatch Resource.
+	maxCepsInCeb int
+}
+
+// Create and Initialize a new FCFS based manager
+func newCebManagerFcfs(cebSync *aggregator, maxCepsInCeb int) cebManager {
+	return &cebManagerFcfs{
+		cacheCeb:     make(map[string]*ceBatch),
+		cebsToSync:   cebSync,
+		cacheCepMap:  newCepToCebMapping(),
+		maxCepsInCeb: maxCepsInCeb,
+	}
+}
+
+// queueCep insert the CEP in a CEB, if the CEP already exists in a CEB
+// it replaces with new CEP.
+func queueCep(cep *cilium_v2.CoreCiliumEndpoint, ceBatch *ceBatch) {
+	// If cep already exists in ceb, compare new cep with cached cep.
+	// Update only if there is any change.
+	for i, ep := range ceBatch.ceb.Endpoints {
+		if ep.Name == cep.Name && ep.Namespace == cep.Namespace {
+			if cep.DeepEqual(&ep) {
+				return
+			}
+			// Remove the matched cep from list.
+			ceBatch.backendMutex.Lock()
```

Comment:
I'm not following this. If we are getting the `ceBatch.backendMutex.Lock()` here, why aren't we getting a `ceBatch.backendMutex.Lock()` to read `ceBatch.ceb.Endpoints` on line 91?

Reply:
The `ceBatch.backendMutex.Lock()/Unlock()` is used only to protect `writes` going into `ceBatch.ceb.Endpoints`.  
There are two threads running, one in the context of `CEP watcher` and another reconciler thread. 
Reconciler thread, gets copy of CEB data before syncing with api-server, hence it need clean copy of CEB.

Reply:
I refactored code to guard against for Read and Write.